Use logical coordinates in DisplayBuffer

as-cii opened the pull request:
This PR moves all the code that handles pixel positions into `TextEditorPresenter`, where it belongs. This is part of the effort to merge #8811, as we realized we needed the `DisplayBuffer` to be clean before proceeding with measurements.

This refactoring was quite big, so I'd :heart: to gather some feedback on the changes I have made (both on code quality and regressions).

What's missing:
- [x] Fix missing failing specs.
- [x] Fix package specs (by shimming the original API).
- [x] Maintain a scroll queue to preserve scroll history.
- [x] Serialize scroll state in terms of logical coordinates.
- [x] Confirm that packages work as expected.
- [x] Confirm that atom core works as expected.
- [x] Shim `TextEditor` APIs to use their `TextEditorElement` counterpart.
- [x] Fix discovered bugs.

Thanks! :bow:

/cc: @nathansobo @atom/feedback

as-cii committed:
wip

as-cii committed:
wip

as-cii committed:
Remove logs

as-cii committed:
Avoid relying on ::scrollTop to calculate logical height

as-cii committed:
Merge branch 'master' into as-display-buffer-logical-coordinates

as-cii committed:
Move scroll margin into TextEditorPresenter

as-cii committed:
:green_heart: Fix TextEditorComponent specs

as-cii committed:
Update scroll top on ::getState

as-cii committed:
Prevent default for mousewheel event on mini editors

(ref. 0346e58)

as-cii committed:
Make sure scroll positions are integers

as-cii committed:
Start porting scroll-related specs

as-cii committed:
Port .selectLinesContainingCursors() specs

as-cii committed:
Port autoscroll when selecting buffer ranges

as-cii committed:
wip

as-cii committed:
Port .scrollToCursorPosition() specs

as-cii committed:
:fire: Remove assertion about scrolling

as-cii committed:
Finish porting TextEditor specs

as-cii committed:
Start porting DisplayBuffer specs

as-cii committed:
:fire: Remove specs

...as we are already testing a similar behavior in the presenter

as-cii committed:
Test that the logical scroll event is triggered

as-cii committed:
:fire: Remove double tested behavior

as-cii committed:
:fire: Remove absolute scrolling from DisplayBuffer

as-cii committed:
Stop reporting scroll positions to the model

as-cii committed:
Report the already calculated height to the model

as-cii committed:
Report the already calculated width to the model

as-cii committed:
Save scroll positions in the model to serialize 'em

as-cii committed:
Fix soft wrapping when editorWidthInChars is used

as-cii committed:
:fire: Remove unused code

as-cii committed:
:green_heart: Fix remaining failures in TextEditorPresenter

as-cii committed:
:green_heart: Fix some TextEditorComponent specs

as-cii committed:
Commit pending positions when a frame is being served

as-cii committed:
Preserve un-rounded scroll positions

as-cii committed:
Continue emitting events when setting scroll top in the model

as-cii committed:
:fire: Remove unused method

as-cii committed:
:fire:

as-cii committed:
Merge branch 'master' into as-display-buffer-logical-coordinates

# Conflicts:
#	src/display-buffer.coffee

as-cii committed:
Bring back the visible row range API

nathansobo commented on src/display-buffer.coffee:
```diff
@@ -324,84 +253,17 @@ class DisplayBuffer extends Model
-        @setScrollRight(desiredScrollRight)
+  scrollToScreenRange: (screenRange, options = {}) ->
+    scrollEvent = {screenRange, options}
+    @emitter.emit "did-change-scroll-position", scrollEvent
```
How would you feel about expressing the position logically as `scrollTopRow`, and `scrollLeftColumn`, rather than ephemerally as an event. That would mean you could assign the row/column to request a scroll.

as-cii replied:
> How would you feel about expressing the position logically as scrollTopRow, and scrollLeftColumn, rather than ephemerally as an event. That would mean you could assign the row/column to request a scroll, and also help with serialization.
> 
> Edit: On further thought, I'm less sure about this. It would require us to pull more view state into the display buffer. Maybe a 1-time request to scroll to a point is better.

That's how it looked like in the first commits: basically, I wanted to express scroll positions in terms of logical coordinates. At some point, however, I realized that scrolling is merely a view concern: in other words, if a view isn't attached to a `DisplayBuffer`, it's kinda weird to _talk_ about scrolling in that object.

And I think the same goes for scoped character widths or other similar concerns that are still in the display buffer: I feel like this is just a step towards removing all of that knowledge from an object whose purpose is to handle spatial coordinates independently of how they get presented to the user.

What do you think?

nathansobo replied:
:+1:

nathansobo commented on src/display-buffer.coffee:
```diff
@@ -1314,6 +1154,25 @@ class DisplayBuffer extends Model
       @emitter.emit 'did-create-marker', marker
 
+  # TODO: serialize state in TextEditorElement, rather than saving scroll
+  # positions here.
```
Imagine if your font is a different size next time you load the editor... don't you actually want the same logical location rather than the same pixel location to be in view? That's what makes me wonder about expressing the scroll top as state in terms of rows... but maybe it's something that is only assigned by the view and read one time when it is attached to the DOM.

as-cii replied:
> but maybe it's something that is only assigned by the view and read one time when it is attached to the DOM.

:+1: for this. Ideally, I'd :heart: the view to have a serialized state of its own because of the reasons pointed out in https://github.com/atom/atom/pull/8905#discussion_r40288745, but I am :cool: with keeping the model up to date with the current logical coordinates as well.

as-cii committed:
Merge branch 'master' into as-display-buffer-logical-coordinates

as-cii committed:
Serialize state using logical coordinates

as-cii committed:
Remove references to TextEditor::scroll...

as-cii committed:
Restore scroll position after we have pixel position requirements

as-cii committed:
Let's start with deprecations!

as-cii committed:
wip

as-cii committed:
:green_heart: Fix specs

as-cii committed:
Adjust pixelRectForScreenRange to return absolute values

as-cii committed:
Port leftover specs from DisplayBuffer

as-cii committed:
Ensure realScroll... is up to date with scroll...

as-cii committed:
Maintain scroll history

as-cii committed:
Revert "Maintain scroll history"

This reverts commit 2e6bb53303175823b55f6ddf347b66a754170cf0.

as-cii committed:
Enforce a 'last scroll wins' model for batched scrolls

as-cii committed:
Prevent default event for actions that won't lead to scrolling

as-cii committed:
:art: Rename to ::onDidRequestAutoscroll

as-cii committed:
Shim ::setWidth and ::getWidth to TextEditorElement

as-cii committed:
Shim ::setHeight and ::getHeight to TextEditorElement

as-cii committed:
:green_heart:

as-cii committed:
:fire: Remove deprecated methods

as-cii committed:
:art: Use ::setWidth and ::setHeight

as-cii committed:
:fire:

as-cii commented:
Okay, the build is :green_heart: now and I think I have addressed all the concerns we had during our call. :tada:

I have found, however, a method on `Marker` which makes use of the display buffer's pixel conversion methods (i.e. `Marker::getPixelRange`) and a cursory search on GitHub showed that someone is actually using it :disappointed:.

We cannot shim it to a corresponding method on the view, though, as we don't have any reference to `TextEditor` in `DisplayBuffer`: passing a `TextEditor` instance when constructing `DisplayBuffer` seems like a bad move to me (other than being an unhealthy dependency, I think we'd have serialization issues) so I went ahead and simply deleted it in a027772 and 7cd318f.

If we want to keep it, I think the only option would be to deprecate it for a few versions and then delete it afterwards (so that package developers have some time to upgrade). On the other hand I'd :heart: to clean the `DisplayBuffer` as soon as possible... What do you think?

Other than that, I think this is ready to :ship:! @nathansobo: please, have one last :eyes:.

Thanks everyone for the feedback! :pray:

nathansobo commented:
> and a cursory search on GitHub showed that someone is actually using it

It wasn't in the public API right? How popular is the package in question? Maybe we should open an issue or even open a PR if it's a simple fix.

nathansobo commented on src/text-editor-presenter.coffee:
```diff
@@ -1438,3 +1474,179 @@ class TextEditorPresenter
+
+    iterator = @model.tokenizedLineForScreenRow(row).getTokenIterator()
+    while iterator.next()
+      charWidths = @getScopedCharacterWidths(iterator.getScopes())
```
It's going to be so great to drop this.

as-cii replied:
Agreed! :fireworks:

nathansobo commented on src/text-editor.coffee:
```diff
@@ -2854,23 +2859,27 @@ class TextEditor extends Model
 
   # Essential: Scrolls the editor to the top
   scrollToTop: ->
-    @setScrollTop(0)
```
Should we drop deprecated methods from the public API to avoid encouraging further use? Ideally we'd have a *Deprecated* label, but I'm not sure it's worth the time to add it to the docs system at this time.

as-cii replied:
Agreed to drop methods from the public API. :+1:

nathansobo commented:
Gave this a quick scan. Ultimately if tests are green and the structure hasn't changed beyond our discussion, I'm good to merge once we drop deprecated methods from the docs.

as-cii committed:
:fire: Drop deprecated methods from the public API

as-cii commented:
> It wasn't in the public API right? How popular is the package in question? Maybe we should open an issue or even open a PR if it's a simple fix.

Nope, the method wasn't in the public API. The package in question is `color-picker` but I think the fix is straightforward enough to deserve a PR. I'll do that shortly and :ship: this PR afterwards.

Thanks, @nathansobo!

as-cii committed:
:art:

as-cii committed:
:fire: Remove unused code

nathansobo commented:
💥🎉🇮🇹

maxbrunsfeld commented:
Awesome!

maxbrunsfeld commented:
Hey @as-cii, apparently our internal CI goes red when deprecated methods are called by core or bundled packages. I actually didn't know this either, until now, but I guess it makes sense, because we don't want the somewhat-expensive `Grim.deprecate` calls to be happening in the next release. I [temporarily commented out](https://github.com/atom/atom/commit/8fb4e887df800438b2deeb23a2c23004b9ff358a) the `Grim.deprecate` calls that you added in `TextEditor`, just to get CI green.

When you get a chance, could you update core and bundled packages to use the new APIs? The deprecations occurred when these tests:
* [ ] find-and-replace
* [ ] symbols-view
* [ ] wrap-guide
* [ ] atom core
  * [ ] `::page{Up,Down}`
  * [ ] `::selectPage{Up,Down}`
